### PR TITLE
feat(route-recognizer): support for route parameter constraints

### LINKED ISSUE #1734: [RFC] Route parameter constraints
## TL;DR

- Proposal of route parameter constraint based on RegExp API and pattern. Example: `'product/:id{{\d+}}'`
- Usage is optional; when present, the route recognizer will respect that.
- Out-of-scope: The usage of wildcard pattern using `*`, `**` etc.
- Limitation: The usage of `/` in the pattern is not allowed.

# 💬 RFC

This RFC proposes application of pattern constraints for route parameter.

## 🔦 Context

The route parameters can be required, optional, or wildcard. So far, there is no other kind of constrains applied on the parameters. However, sometimes it might be desirable to restrict the navigation only if the given parameters meet certain patterns. This RFC proposes an extension in the route parameters recognition based on optional regex pattern. If no pattern is specified, then the route recognizer works as it works now. However, if a constraint is specified for a route parameter, then the route recognizer only matches/recognizes the given route, if and only if all the constraints specified for all the route parameters are fulfilled.

This is explained further using the examples below.

## 💻 Examples

Restrict the `id` parameter to the `product` route only to integer.

```typescript
'product/:id{{\d+}}'
```

The `{{` and `}}` are the **terminal tokens** that indicate the start and the end of the pattern. AFAIK, that should not pose a conflict to the possible regex syntaxes. The [URLPattern API](https://developer.mozilla.org/en-US/docs/Web/API/URL_Pattern_API) uses `(`, and `)` as terminal tokens. However, I think that adopting a non-conflicting token will keep the parsing logic easy and also make scope for more intricate regex stuff in future; for example, capture groups.

Let us consider another example. The following example, restricts the `tab` parameter to 'items', 'payment', and 'delivery'.

```typescript
'order/:tab?{{items|payment|delivery}}'
```

Note that we don't want to parse the regex by ourselves, and rather use the **standard RegExp API** for that. Hence, what is supported by the RegExp API should also be supported by this.

This also means that glob/wildcard patterns like below will not be supported and are **out of the scope** of the change proposed by this RFC.

```typescript
'view-file/:name?{{*.txt}}' // out of the scope
```

However, one can easily write a regex for that; a naive approach may look like below.

```typescript
'view-file/:name?{{^.+\.txt$}}'
```

**One limitation** is the usage of `/` in the parameter constraint. Because in the context of URL, that is a path separator and a parameter value can never contain that.

## Related works

- [ASP.NET Web API route constraints](https://learn.microsoft.com/en-us/aspnet/web-api/overview/web-api-routing-and-actions/attribute-routing-in-web-api-2#route-constraints).
- [URLPattern API](https://developer.mozilla.org/en-US/docs/Web/API/URL_Pattern_API).

## Future extension possibilities (currently out of the scope of this proposal)

- A more generic syntax: `'product/:id{{$contraint:regex:\d+}}'`. 
- Out of the box patterns: `{{$constraint:int}}`, `{{$constraint:range:10:42}}` etc.
- Custom constraint: `{{$constraint:notInBlacklist:foo:bar}}`.
- `IConstraint`

  ```ts
  interface IConstraint<T = unknown> {
    hydrate(raw: string): void;
    match(raw: string): boolean;
    coerce(raw: string): T;
  }
  ```
- Constraints metadata can be registered and resolved, enabling constraint discovery by name: `@routeConstriant('notInBlacklist')`.
- Type coercion according to the constraints.

### PR DESCRIPTION
**Pull request by Sayan751**
# Pull Request

## 📖 Description

This PR adds the support for route parameter constraints as described in the RFC #1734.

### 🎫 Issues

Closes #1734

## ⏭ Next Steps

- Update docs.
- Enhance feature as outlined in #1734.

### CODE REVIEW

**Review by bigopon — COMMENTED** [review]
Nice work, I have only a question for now.

File: packages/route-recognizer/src/index.ts [hunk 1]
```diff
@@ -395,11 +428,21 @@ export class RouteRecognizer<T> {
...
+          const match = routeParameterPattern.exec(part);
+          const { name, optional } = match?.groups ?? {};
+          const isOptional = optional === '?';
           if (name === RESIDUE) throw new Error(`Invalid parameter name; usage of the reserved parameter name '${RESIDUE}' is used.`);
-          params.push(new Parameter(name, isOptional, false));
-          state = new DynamicSegment<T>(name, isOptional).appendTo(state);
+          let pattern: RegExp | null = null;
+          let constraint = match?.groups?.constraint;
+          if (constraint != null) {
+            if (!constraint.startsWith('^') && !constraint.endsWith('$')) {
```
[the remark this answers is not in the document]
Comment (Sayan751) [reply packages/route-recognizer/src/index.ts:439]: I agree with you that it is not entirely WYSIWYG. I still think that I need to mention a couple of points.

Firstly, the end marker of `^` or `$` is not for the complete route and rather for a path segment. More accurately, for a single route parameter. For example, if we have a route like `a/:id`, we can constrain the possible values of `:id`, using this feature. If I have in any way misunderstood your example, please feel free to elaborate.

Secondly, I have construed the following to be a common example/use-case of this feature.

```ts
a/:tab{{foo|bar}}
``` 

I think the common expectation is that the routes such as `a/foo` or `a/bar` are recognized, and the routes such as `a/foobar`, `a/barfoo`, or `a/foowhatever` are not recognized. The end markers only ensure that the complete value of the parameter satisfies the pattern, when the user has omitted the end markers. You can say that user can add the end markers and if you think that is rather the common expectation, I can remove adding those markers.

[the remark this answers is not in the document]
Comment (bigopon) [reply packages/route-recognizer/src/index.ts:439]: I think we should leave that decision to devs. Without automatically added anchors, devs can express both normal and anchorred patterns. Its not possible with automatic anchor.

[the remark this answers is not in the document]
Comment (Sayan751) [reply packages/route-recognizer/src/index.ts:439]: I agree. I will remove this now. Probably later, we can come up with some predefined convenience pattern like `int`, `enum` etc.

[the remark this answers is not in the document]
Comment (Sayan751) [reply packages/route-recognizer/src/index.ts]: It is done now.

**Review by bigopon — APPROVED** [review]
Nice work, lgtm :+1:

**Pull request merged**